### PR TITLE
New methods for Lifecycle: catchAll + foldM

### PR DESCRIPTION
**Pull request by VladPodilnyk**
Seems OK.
Feel free to throw tomatoes at me xD
Partially addresses https://github.com/7mind/izumi/issues/1309

### CODE REVIEW

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala [hunk 1]
```diff
@@ -240,6 +240,8 @@ trait Lifecycle[+F[_], +OuterResource] {
   final def evalMap[G[x] >: F[x]: QuasiIO, B](f: OuterResource => G[B]): Lifecycle[G, B] = evalMapImpl[G, OuterResource, B](this)(f)
   final def evalTap[G[x] >: F[x]: QuasiIO](f: OuterResource => G[Unit]): Lifecycle[G, OuterResource] =
     evalMap[G, OuterResource](a => QuasiIO[G].map(f(a))(_ => a))
+  final def foldM[G[x] >: F[x]: QuasiIO, B](onFailure: Throwable => Lifecycle[G, B], onSuccess: OuterResource => Lifecycle[G, B]): Lifecycle[G, B] =
```
Comment (neko-kai) [inline distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: Rename just to be consistent with `izumi.functional.bio.Error3#redeem`:
```suggestion
  final def redeem[G[x] >: F[x]: QuasiIO, B](onFailure: Throwable => Lifecycle[G, B], onSuccess: OuterResource => Lifecycle[G, B]): Lifecycle[G, B] =
```

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala [hunk 2]
```diff
@@ -962,6 +967,61 @@ object Lifecycle extends LifecycleCatsInstances {
...
+          use = use
+        )
+      }
+
+      override def acquire: F[InnerResource] = F.maybeSuspend(new AtomicReference(Nil))
+      override def release(finalizers: InnerResource): F[Unit] = F.suspendF(F.traverse_(finalizers.get())(_.apply()))
+      override def extract[C >: B](finalizers: InnerResource): Either[F[C], C] = Left {
+        bracketAppendFinalizer(finalizers)(self) {
+          inner1: self.InnerResource =>
+            F.definitelyRecover {
```
Comment (neko-kai) [inline distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: This will not run the failure handler if `self.acquire` failed. Current `Lifecycle` encoding is roughly similar to [old pre-RC19 ZManaged encoding](https://github.com/zio/zio/blob/v1.0.0-RC18/core/shared/src/main/scala/zio/ZManaged.scala) (where `Lifecycle#extract` = `Reservation#acquire`, `Lifecycle#acquire` = `ZManaged#reservation`), you may be able to look at their implementation of [foldM](https://github.com/zio/zio/blob/v1.0.0-RC18/core/shared/src/main/scala/zio/ZManaged.scala#L396) for tips.

Comment (VladPodilnyk) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: Yeah, I've taken inspiration there, but seems I misunderstood some parts. Thanks for the review and you notes about implementation!

Comment (VladPodilnyk) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: Done.

**Comment by VladPodilnyk** [comment]
I guess, next step is to make Lifecycle with F[_, _] which gives ability to work with different error types.
This seems like major change or not ?

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala [hunk 3]
```diff
@@ -962,6 +967,61 @@ object Lifecycle extends LifecycleCatsInstances {
...
+
+      override def acquire: F[InnerResource] = F.maybeSuspend(new AtomicReference(Nil))
+      override def release(finalizers: InnerResource): F[Unit] = F.suspendF(F.traverse_(finalizers.get())(_.apply()))
+      override def extract[C >: B](finalizers: InnerResource): Either[F[C], C] = Left {
+        F.definitelyRecover {
+          bracketAppendFinalizer(finalizers)(self) {
+            inner1: self.InnerResource =>
+              self.extract(inner1).fold(_.map(success), F pure success(_)).flatMap {
+                that: Lifecycle[F, B] =>
+                  bracketAppendFinalizer(finalizers)(that) {
```
Comment (neko-kai) [inline distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: @VladPodilnyk It is possible for `failure` and `success` to both run at the same time with this scheme, if the failure occurs at or after this line, which is logically incorrect.

Comment (VladPodilnyk) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: @neko-kai, now should be fixed. I have some issues with Lifecycle instance for Identity. It does not work now, since action is eagerly evaluated.

File: distage/distage-core-api/src/main/scala/izumi/distage/model/effect/QuasiIO.scala [hunk 4]
```diff
@@ -51,6 +51,8 @@ trait QuasiIO[F[_]] extends QuasiApplicative[F] {
     */
   def definitelyRecoverCause[A](action: => F[A])(recoverCause: (Throwable, (() => Throwable)) => F[A]): F[A]
 
+  def redeem[A, B](action: F[A])(failure: Throwable => F[B], success: A => F[B]): F[B]
```
Comment (neko-kai) [inline distage/distage-core-api/src/main/scala/izumi/distage/model/effect/QuasiIO.scala]: @VladPodilnyk Are you sure a separate `redeem` method is required when it's equivalent to `action.definitelyRecover(failure).flatMap(success)`?

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala [hunk 5]
```diff
@@ -962,6 +967,64 @@ object Lifecycle extends LifecycleCatsInstances {
...
+      override def acquire: F[InnerResource] = F.maybeSuspend(new AtomicReference(Nil))
+      override def release(finalizers: InnerResource): F[Unit] = F.suspendF(F.traverse_(finalizers.get())(_.apply()))
+      override def extract[C >: B](finalizers: InnerResource): Either[F[C], C] = Left {
+        val action = bracketAppendFinalizer(finalizers)(self) {
+          inner: self.InnerResource =>
+            self.extract(inner).fold(identity, F.pure)
+        }
+
+        val onSuccess = (a: A) => {
+          val resource = success(a)
```
Comment (neko-kai) [inline distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:1008]: This extraction into a val cannot be done safely due to `Identity` instance

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala [hunk 6]
```diff
@@ -962,6 +967,64 @@ object Lifecycle extends LifecycleCatsInstances {
...
+    self: Lifecycle[F, A]
+  )(failure: Throwable => Lifecycle[F, B],
+    success: A => Lifecycle[F, B],
+  )(implicit F: QuasiIO[F]
+  ): Lifecycle[F, B] = {
+    import QuasiIO.syntax._
+    new Lifecycle[F, B] {
+      override type InnerResource = AtomicReference[List[() => F[Unit]]]
+
+      private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
```
Comment (neko-kai) [inline distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:981]:
```suggestion
      private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: => Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
```

Must suspend here to ensure that any effects in construction of lifecycle are wrapped into `bracket`'s `try` (in Identity instance). (Lifecycle/QuasiIO must work correctly with impure code)

Comment (neko-kai) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:981]: Hmm, this may be tricky because of path dependency

File: distage/distage-core-api/src/main/scala/izumi/distage/model/effect/QuasiIO.scala (hunk 4 above)
Comment (neko-kai) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/effect/QuasiIO.scala]:
```suggestion
  def redeem[A, B](action: => F[A])(failure: Throwable => F[B], success: A => F[B]): F[B]
```
Also you would need suspension here to make it work with `Identity`.

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala [hunk 7]
```diff
@@ -962,6 +967,64 @@ object Lifecycle extends LifecycleCatsInstances {
...
+          }
+        )(release = _ => F.unit)(
+          use = use
+        )
+      }
+
+      override def acquire: F[InnerResource] = F.maybeSuspend(new AtomicReference(Nil))
+      override def release(finalizers: InnerResource): F[Unit] = F.suspendF(F.traverse_(finalizers.get())(_.apply()))
+      override def extract[C >: B](finalizers: InnerResource): Either[F[C], C] = Left {
+        val action = bracketAppendFinalizer(finalizers)(self) {
```
Comment (neko-kai) [inline distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: `val action` must be inlined or become `def action`, because with `F=Identity` it's not an `action` but a `result`

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala (hunk 6 above)
Comment (neko-kai) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:981]: @VladPodilnyk 
This seems to work as an encoding that supports suspension and path dependency:
```scala
private[this] sealed trait LifecyclePair[+a] {
  val lifecycle: Lifecycle[F, a]
  val inner: lifecycle.InnerResource
}
private[this] def LifecyclePair[a](lifecycle0: Lifecycle[F, a])(inner0: lifecycle0.InnerResource): LifecyclePair[a] = new LifecyclePair[a] {
  override val lifecycle: lifecycle0.type = lifecycle0
  override val inner: lifecycle.InnerResource = inner0
}

private[this] def bracketAppendFinalizer[a, b, t](
  finalizers: InnerResource
)(lifecycleCtor: () => Lifecycle[F, a]
)(use: LifecyclePair[a] => F[b]
): F[b] = {
  F.bracket(
    acquire = {
      val lifecycle = lifecycleCtor()
      lifecycle.acquire.flatMap {
        a =>
          F.maybeSuspend {
            // can't use `.updateAndGet` because of Scala.js
            var oldValue = finalizers.get()
            while (!finalizers.compareAndSet(oldValue, (() => lifecycle.release(a)) :: oldValue)) {
              oldValue = finalizers.get()
            }
            LifecyclePair(lifecycle)(a)
          }
      }
    }
  )(release = _ => F.unit)(
    use = use
  )
}
```

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala (hunk 3 above)
Comment (neko-kai) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: @VladPodilnyk I think I don't see any issues now, seems like the only thing left is a few by-names to accomodate `Identity`

File: distage/distage-core-api/src/main/scala/izumi/distage/model/effect/QuasiIO.scala (hunk 4 above)
Comment (VladPodilnyk) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/effect/QuasiIO.scala]: >@VladPodilnyk Are you sure a separate redeem method is required when it's equivalent to action.definitelyRecover(failure).flatMap(success)?

The problem is action has type `F[A]`, failure is a function from `Throwable to F[B]` hence I couldn't use `definatelyRecover` since it requires the same type for action and failure. So I decided to add some functionality.

Comment (neko-kai) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/effect/QuasiIO.scala]: Right, I wasn't correct here

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala (hunk 6 above)
Comment (VladPodilnyk) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:981]: Do you propose to extract it into a pattern ☝️ ?
(it's not clear for me what problem does it solves)
Also, we probably should come up with another name bc this one is not very informative, IMO.

Comment (neko-kai) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:981]: @VladPodilnyk 
> (it's not clear for me what problem does it solves)

You need the `lifecycle` parameter to be by-name because it may produce effects immediately when called, instead of when executed (it's natural to perform effects inplace if you use impure Identity instance)

But if you write it as by-name, it's now impossible to write the `lifecycle.InnerResource` type:
```scala
private[this] def bracketAppendFinalizer[a, b](finalizers: InnerResource)(lifecycle: => Lifecycle[F, a])(use: lifecycle.InnerResource => F[b]): F[b] = {
```

This type will produce an error because `lifecycle` is now by-name - every time you refer to it in code, the by-name parameter will run again and produce a new value (operationally `=> Lifecycle[F, a]` is exactly the same as `() => Lifecycle[F, a]`). 
Because it produces a new lifecycle each time, you cannot refer to `lifecycle.InnerResource` — a type member of `lifecycle` value — because there is no "`lifecycle` value" here, only a function that makes new Lifecycles. So we package the result "value" into a trait to make it referable (the reason it's not a class is because classes do not support path-dependency in Scala 2, only traits)

In short, try to implement the above signature and you'll be able to experience the limitations that force it to be this way.

Comment (neko-kai) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:981]: Right now you need to get rid of `val` extraction and write this thing as if everything is imperative (because it IS when using `Identity`) and then it's good to merge 🚀

Comment (VladPodilnyk) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:981]: @neko-kai, thanks for this excellent explanation!

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala (hunk 5 above)
Comment (VladPodilnyk) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala:1008]: Replaced with `def` now.

File: distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala (hunk 7 above)
Comment (VladPodilnyk) [reply distage/distage-core-api/src/main/scala/izumi/distage/model/definition/Lifecycle.scala]: Done.

**Review by neko-kai — APPROVED** [review]

**Comment by neko-kai** [comment]
@VladPodilnyk Thank you! 🙏

**Comment by VladPodilnyk** [comment]
:hugs:
